sql-parser/src/main/java/io/crate/sql/tree/Insert.java

```diff
@@ -21,20 +21,23 @@
 
 package io.crate.sql.tree;
 
+import com.google.common.base.MoreObjects;
 import com.google.common.base.Objects;
 
 import java.util.Collections;
 import java.util.List;
 
-public abstract class Insert<T> extends Statement {
+public final class Insert<T> extends Statement {
 
-    protected final Table<T> table;
+    private final Table<T> table;
     private final DuplicateKeyContext<T> duplicateKeyContext;
-    protected final List<String> columns;
+    private final List<String> columns;
+    private final Query subQuery;
 
-    Insert(Table<T> table, List<String> columns, DuplicateKeyContext<T> duplicateKeyContext) {
+    public Insert(Table<T> table, Query subQuery, List<String> columns, DuplicateKeyContext<T> duplicateKeyContext) {
         this.table = table;
         this.columns = columns;
+        this.subQuery = subQuery;
```

Comment:
maybe `insertSource`, just `query` or smth else? Because it can a bit confusing when you have values in there, but the variable is called `subQuery`.

Otherwise, lgtm